core/state/pruner/pruner.go

```diff
@@ -0,0 +1,511 @@
+	// - Neither these `prune-state` running is finished(e.g. interrupted manually)
+	// - The state bloom filter is already generated, a part of state is deleted,
+	//   so that resuming the pruning here is mandatory
+	// - The state HEAD is rewound already because of multiple incomplete `prune-state`
+	// In this case, even the state HEAD is not exactly matched with snapshot, it
+	// still feasible to recover the pruning correctly.
+	snaptree, err := snapshot.New(db, trie.NewDatabase(db), 256, headHeader.Root, false, false, true)
+	if err != nil {
+		return err // The relevant snapshot(s) might not exist
+	}
+	stateBloom, err := NewStateBloomFromDisk(stateBloomPath)
+	if err != nil {
+		return err
+	}
+	log.Info("Loaded the state bloom filter", "path", stateBloomPath)
+
+	// Before start the pruning, delete the clean trie cache first.
+	// It's necessary otherwise in the next restart we will hit the
+	// deleted state root in the "clean cache" so that the incomplete
+	// state is picked for usage.
+	deleteCleanTrieCache(trieCachePath)
+
+	if err := prune(db, stateBloom, time.Now()); err != nil {
+		return err
+	}
+	// Pruning is done, now drop the "useless" layers from the snapshot.
+	// Firstly, flushing the target layer into the disk. After that all
+	// diff layers below the target will all be merged into the disk.
+	if err := snaptree.Cap(stateBloomRoot, 0); err != nil {
+		return err
+	}
+	// Secondly, flushing the snapshot journal into the disk. All diff
+	// layers upon are dropped silently. Eventually the entire snapshot
+	// tree is converted into a single disk layer with the pruning target
+	// as the root.
+	if _, err := snaptree.Journal(stateBloomRoot); err != nil {
+		return err
+	}
+	// Delete the state bloom, it marks the entire pruning procedure is
+	// finished. If any crashes or manual exit happens before this,
+	// `RecoverPruning` will pick it up in the next restarts to redo all
+	// the things.
+	os.RemoveAll(stateBloomPath)
+	return nil
+}
+
+// extractGenesis loads the genesis state and creates the nodes marker.
+// So that it can be used as an present indicator for all genesis trie nodes.
+func extractGenesis(db ethdb.Database) (map[common.Hash]struct{}, error) {
+	genesisHash := rawdb.ReadCanonicalHash(db, 0)
+	if genesisHash == (common.Hash{}) {
+		return nil, errors.New("missing genesis hash")
+	}
+	genesis := rawdb.ReadBlock(db, genesisHash, 0)
+	if genesis == nil {
+		return nil, errors.New("missing genesis block")
+	}
+	t, err := trie.New(genesis.Root(), trie.NewDatabase(db))
+	if err != nil {
+		return nil, err
+	}
+	marker := make(map[common.Hash]struct{})
+	accIter := t.NodeIterator(nil)
+	for accIter.Next(true) {
+		node := accIter.Hash()
+
+		// Embedded nodes don't have hash.
+		if node != (common.Hash{}) {
+			marker[node] = struct{}{}
+		}
+		// If it's a leaf node, yes we are touching an account,
+		// dig into the storage trie further.
+		if accIter.Leaf() {
+			var acc struct {
```

Comment:
Please use `state.Account`, I think the `state` package does not use the `pruner` package, so it should be fine.